compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs

```diff
@@ -145,6 +148,27 @@ kindOf = \case
   TStruct recordType -> checkRecordType recordType $> KStar
   TNat _ -> pure KNat
 
+expandTypeSynonyms :: MonadGamma m => Type -> m Type
+expandTypeSynonyms = expand where
+  expand = \case
+    TVar v -> return $ TVar v
+    TSyn tsyn -> do
+      DefTypeSyn{synType} <- inWorld (lookupTypeSyn tsyn)
+      expandTypeSynonyms synType
+    TCon tcon -> return $ TCon tcon
+    TApp tfun targ -> do
+      tfun' <- expand tfun
+      targ' <- expand targ
+      return $ TApp tfun' targ'
+    TBuiltin btype -> return $ TBuiltin btype
+    TForall (v, k) t1 -> do
+      t1' <- expand t1
```

Comment:
Note that `v` can be used as a parameter to a type synonym in `T1` so you might end up having to track variables that are in scope when you add substitution.

Reply:
I'm pretty sure that because there are no free vars in the body of the type-syn definition, then we should be fine to just use `substitute`

Reply:
:+1:

Reply:
Turns out we did need to track the variable `v` !